dali/kernels/transpose/transpose_gpu_setup.cuh

```diff
@@ -0,0 +1,211 @@
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TRANSPOSE_TRANSPOSE_GPU_SETUP_CUH_
+#define DALI_KERNELS_TRANSPOSE_TRANSPOSE_GPU_SETUP_CUH_
+
+#include <cuda_runtime.h>
+#include <utility>
+#include "dali/core/tensor_view.h"
+#include "dali/core/fast_div.h"
+#include "dali/kernels/transpose/transpose_gpu_impl.cuh"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace kernels {
+namespace transpose_impl {
+
+template <size_t static_size>
+inline void CollapseUnitDims(TensorShape<> &shape, SmallVector<int, static_size> &perm) {
+  int ndim = shape.size();
+  SmallVector<int, static_size> dim_map, out_perm;
+  TensorShape<> out_shape;
+  dim_map.resize(ndim, -1);
+  for (int i = 0; i < ndim; i++) {
+    if (shape[i] != 1) {
+      dim_map[i] = out_shape.size();
+      out_shape.shape.push_back(shape[i]);
+    }
+  }
+  for (int i = 0; i < ndim; i++) {
+    int m = dim_map[perm[i]];
+    if (m >= 0)
+      out_perm.push_back(m);
+  }
+  shape = std::move(out_shape);
+  perm = std::move(out_perm);
+}
+
+template <size_t static_size>
+inline void CollapseAdjacentDims(TensorShape<> &shape, SmallVector<int, static_size> &perm) {
+  int ndim = shape.size();
+
+  SmallVector<int, static_size> dim_map, inv_perm, out_perm;
+  TensorShape<> out_shape;
+
+  dim_map.resize(ndim, -1);
+
+  inv_perm.resize(ndim);
+  for (int i = 0; i < ndim; i++)
+    inv_perm[perm[i]] = i;
+
+  int nkeep = 0;
+  int64_t extent = 1;
+  int last = 0;
+  for (int i = 0; i < ndim; i++) {
+    if (inv_perm[i] != inv_perm[last] + i - last) {
+      dim_map[i] = nkeep++;
+      out_shape.shape.push_back(extent);
+      last = i;
+      extent = shape[i];
+    } else {
+      extent *= shape[i];
+    }
+    if (nkeep == 0) {
+      dim_map[i] = nkeep++;
+    }
+  }
+
+  out_shape.shape.push_back(extent);
+
+  for (int i = 0; i < ndim; i++) {
+    int pmap = dim_map[perm[i]];
+    if (pmap >= 0)
+      out_perm.push_back(pmap);
+  }
+
+  shape = std::move(out_shape);
+  perm = std::move(out_perm);
+}
+
+template <size_t static_size>
+inline void SimplifyPermute(
+    TensorShape<> &out_shape, SmallVector<int, static_size> &out_perm,
+    const int64_t *shape, const int *perm, int ndim) {
+  out_shape = { shape, shape + ndim };
+  out_perm = { perm, perm + ndim };
+
+  CollapseUnitDims(out_shape, out_perm);
+  CollapseAdjacentDims(out_shape, out_perm);
+}
+
+template <typename T>
+void InitTiledTranspose(TiledTransposeDesc<T> &desc,
+                        const TensorShape<> &shape, span<const int> perm,
+                        same_as_t<T> *out = nullptr, same_as_t<const T> *in = nullptr,
+                        int grid_size = 1) {
+  int ndim = shape.size();
+
+  CalcStrides(desc.in_strides, shape);
+
+  TensorShape<> out_shape = Permute(shape, perm);
+  TensorShape<> tmp_strides;
+  CalcStrides(tmp_strides, out_shape);
+  for (int i = 0; i < ndim; i++) {
+    desc.shape[i] = shape[i];
+    desc.out_strides[perm[i]] = tmp_strides[i];
```

Comment:
So the output shape is already permuted byt strides are not?

Reply:
The permutation is actually _defined_ by strides. I can explain it in comments.

Reply:
Please add comments.

Reply:
Added on top of the file.